README.md
# S32021
Register based virtual machine (vm) implemented in Python 3.

# Usage
    python fhvm.py <filename>

## Optional arguments
To specify the amount of virtual memory, use
    --registers N

Maximum program size
    --maxprogram N
    
# Opcodes
X - value

P - program

R - register

## I/O
Load integer X into register R

    LOAD X R

Prints the value stored in R

    PRINT R
    
Load integer X into register pointed by R

    LOADP X R

Prints the value stored in register pointed by R

    PRINTP R

Inputs integer into R

    INPUT R

Inputs integer to register pointed by R

    INPUTP R

Copies the value from R1 to R2

    COPY R1 R2


## Arithmetic

Increments R by 1

    INC R

Stores R1+R2 into R3

    ADD R1 R2 R3

Stores R1-R2 into R3

    SUB R1 R2 R3

Stores R1*R2 into R3

    MUL R1 R2 R3

Stores R1//R2 into R3

    DIV R1 R2 R3

## Jumps
Sets program pointer to P if R1 > R2

    GOG P R1 R2

Sets program pointer to P if R1 < R2

    GOL P R1 R2
    
Sets program pointer to P if R1 = R2

    GOE P R1 R2
 
Sets program pointer to P

    GOTO P

# Halt & other

Halts the program

    HALT

Modifies the program instruction to X

    META P X

Does nothing

    NOP


s3/CHANGELOG.md
# Change Log

All notable changes to the "s3" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

fhvm.py
"""
'Full House' Virtual Machine

Summer School of Science
Project: Turing Complete
Višnjan, 2021
"""
from opcodes import *
from preprocessor import Preprocessor, PreprocessorFunct
import argparse


DEBUG = False


class VirtualMachine:
    def __init__(self, n_registers, p_registers):
        self.registers = n_registers * [0]
        self.n_registers = n_registers

        self.program = p_registers * [0]
        self.p_registers = p_registers
        # Pokazivač na programski registar
        self.i = 0
        self.endpointsOuter = {}


    def __repr__(self):
        return f"[VM] R: {self.registers}\nP:{self.program}"

    # instrukcije
    ################################

    def LOAD(self, value, r):
        self.registers[r] = value

    def LOADP(self, value, r):
        """Loads value into location pointed by r"""
        ptr = self.registers[r]
        self.registers[ptr] = value

    def COPY(self, r1, r2):
        """Copies r1 into r2"""
        self.registers[r2] = self.registers[r1]

    def META(self, p, x):
        """Writes instruction X into p"""
        self.program[p] = x

    def ADD(self, r1, r2, r3):
        """Stores registers[r1]+registers[r2] into registers[r3] """
        self.registers[r3] = self.registers[r1] + self.registers[r2]

    def SUB(self, r1, r2, r3):
        """Stores registers[r1]-registers[r2] into registers[r3] """
        self.registers[r3] = self.registers[r1] - self.registers[r2]

    def MUL(self, r1, r2, r3):
        self.registers[r3] = self.registers[r1] * self.registers[r2]

    def DIV(self, r1, r2, r3):
        self.registers[r3] = self.registers[r1] // self.registers[r2]

    def GOG(self, p, r1, r2):
        """If r1 > r2 GOTO p"""
        if self.registers[r1] > self.registers[r2]:
            self.i = p
            self.i -= 1

    def GOE(self, p, r1, r2):
        """If r1 == r2 GOTO p"""
        if self.registers[r1] == self.registers[r2]:
            self.i = p
            # run loop increments automatically so we subtract 1
            self.i -= 1

    def GOTO(self, p):
        """Unconditional GOTO p"""
        self.i = p
        # run loop increments automatically so we subtract 1
        self.i -= 1

    def INC(self, r):
        """Increments register r"""
        self.registers[r] += 1

    def PRINT(self, r):
        """Prints the value of register r"""
        print(">>>", self.registers[r])

    def CALL(self, rawNewProgram, ): #TODO dovrsi ig

        f = VirtualMachine(self.n_registers, self.p_registers)
        preprocessor = PreprocessorFunct()
        newProgram, fDEFS, endpointsouter = preprocessor.process(rawNewProgram)
        f.preprocess(newProgram)
        f.loadProgram(newProgram)

    ###################
    def loadProgramFile(self, filename):
        preprocessor = Preprocessor()
        program, DEFS, self.endpointsOuter = preprocessor.process(filename)
        print(f'PROGRAM: {program}\n>>======================================================\nDEFS: {DEFS}\n>>======================================================\nEndpointsOuter: {self.endpointsOuter} ')
        
        self.loadProgram(program)

    def loadProgram(self, program):
        if len(program) > self.p_registers:
            print("Error: Program too big!")
            return None
        i = 0
        for p in program:
            self.program[i] = p
            i += 1

    def run(self):
        """Runs the current program"""
        print("===RUN===")
        while True:
            # Trenutna instrukcija ili argument
            code = self.program[self.i]
            if code == HALT:
                break
            elif code == LOAD:
                self.i += 1
                value = self.program[self.i]
                self.i += 1
                r = self.program[self.i]
                self.LOAD(value, r)

            elif code == LOADP:
                self.i += 1
                value = self.program[self.i]
                self.i += 1
                r = self.program[self.i]
                self.LOAD(value, self.registers[r])

            elif code == PRINT:
                self.i += 1
                r = self.program[self.i]
                self.PRINT(r)

            elif code == INPUT:
                self.i += 1
                r = self.program[self.i]
                value = int(input("<<< "))
                self.LOAD(value, r)

            elif code == INPUTP:
                self.i += 1
                r = self.program[self.i]
                value = int(input("<<< "))
                self.LOAD(value, self.registers[r])

            elif code == COPY:
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.COPY(r1, r2)

            elif code == INC:
                self.i += 1
                r = self.program[self.i]
                self.INC(r)

            elif code == ADD:
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.i += 1
                r3 = self.program[self.i]
                self.ADD(r1, r2, r3)

            elif code == SUB:
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.i += 1
                r3 = self.program[self.i]
                self.SUB(r1, r2, r3)

            elif code == MUL:
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.i += 1
                r3 = self.program[self.i]
                self.MUL(r1, r2, r3)

            elif code == DIV:
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.i += 1
                r3 = self.program[self.i]
                self.DIV(r1, r2, r3)

            elif code == GOG:
                self.i += 1
                p = self.program[self.i]
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.GOG(p, r1, r2)

            elif code == GOL:
                self.i += 1
                p = self.program[self.i]
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.GOG(p, r2, r1)

            elif code == GOE:
                self.i += 1
                p = self.program[self.i]
                self.i += 1
                r1 = self.program[self.i]
                self.i += 1
                r2 = self.program[self.i]
                self.GOE(p, r1, r2)

            elif code == GOTO:
                self.i += 1
                p = self.program[self.i]
                self.GOTO(p)

            elif code == META:
                self.i += 1
                p = self.program[self.i]
                self.i += 1
                x = self.program[self.i]
                self.META(p, x)

            elif code == NOP:
                # do nothing
                self.i += 1

            elif code == DEF:
                self.i = self.endpointsOuter[self.i]
            
            elif(code == LIST):
                pass

            else:
                print(f"Error! Unknown instruction: '{code}'")
                break

            # Automatically goes to the next instruction
            self.i += 1
        print("===END===")


if __name__ == "__main__":

    parser = argparse.ArgumentParser(description="Virtual Machine")
    parser.add_argument("file", type=str,
                        help="Name of code file (.s3)")

    parser.add_argument("--registers", type=int, default=64,
                        help="Number of registers")
    parser.add_argument("--maxprogram", type=int,
                        default=64, help="Maximum program size")
    args = parser.parse_args()

    vm = VirtualMachine(args.registers, args.maxprogram)
    vm.loadProgramFile(args.file)
    vm.run()


preprocessor.py
#LOADPROGRAMFILE
#PREPROCESSDEF
#PREPROCESSLINES
#COPYDEF
from opcodes import *


class Preprocessor:
    def __init__(self):
        pass

    def process(self, filename):
        """
        Processes code file into virtual machine instructions
        returns:
            program, DEFS, endpoints
        
        program - list of machine instructions
        DEFS - dictionary with functions and their "names" (order of declaration)
        endpoints - begin and end locations for function declarations

        """
        with open(filename, "r") as file:
            lines = file.readlines()
    
        lines = [line.split(";")[0].split() for line in lines]
        self.argsUpdate(lines)
        rawProgram, lineSet = self.initialRead(lines)
        endpointsOuter, numDefs, endpointsDict = self.preprocessDef(lines, lineSet)
        DEFS = self.copyDef(endpointsOuter,  numDefs, rawProgram)
        program = self.processLines(rawProgram, lines, lineSet)
        print(ARGS)
        return program, DEFS, endpointsDict


    def initialRead(self, lines):
        """
        Returns rawProgram which is not readable to virtual machine
        Turn variables into register locations and commands indo opcodes
        Remembers where lines start in the program instructions
        """
        vars = {}
        rawProgram = []
        lineSet = set()
        i = 0
        for line in lines:
            cmd = line[0]
            lineSet.add(i)
            rawProgram.append(OPCODES[cmd])
            if cmd == 'LIST':
                for h in range(int(line[2])):
                    vars[line[1]+str(h)] = len(vars)
                continue
            for arg in line[1:]:
                if arg < 'A':
                    rawProgram.append(int(arg))
                elif "[" in arg:
                    var = ""
                    for h in arg:
                        name = ""
                        for k in arg:
                            if k == "[":
                                break
                            name += k
                        if (h in [str(g) for g in range(10)]):
                            var += h
                    rawProgram.append(vars[name+var])


                elif arg >= 'A' and arg <= 'Z' or arg >= 'a' and arg <= 'z':
                    if arg not in vars:
                        vars[arg] = len(vars)
                        rawProgram.append(vars[arg])
                    else:
                        rawProgram.append(vars[arg])
                else:
                    print("Error: Incorrect Input Value")
                i += 1
            i += 1
        return rawProgram, lineSet 

    def processLines(self, rawProgram, lines, lineSet):
        """
        For commands that use line pointers 

        Turns arg from line to program location
        """
        i = 0
        proList = sorted(list(lineSet))
        for line in lines:
            cmd = line[0]
            if OPCODES[cmd] in [GOG, GOL, GOE, GOTO]:
                rawProgram[i+1] = proList[int(line[1])-1]
            if OPCODES[cmd] == DEF:
                i += ARGS[(OPCODES[cmd], line[1])]
            elif OPCODES[cmd] == CALL:
                i += ARGS[(OPCODES[cmd], line[1])]
            else:
                i += ARGS[OPCODES[cmd]]
            i += 1
        return rawProgram

    def preprocessDef(self, lines, lineSet):
        """
        Finds locations in the program where declerations of functions begin and end.
        Counts the number of functions 
        Returns list of endpoints,  numDefs
        """
        defs = []
        enddefs = 0
        endpoints = []
        proList = sorted(list(lineSet))
        for i, line in enumerate(lines):
            if line[0] == "DEF":
                defs.append(i)
            elif line[0] == "ENDDEF":
                enddefs = i
                endpoints.append([defs.pop(), enddefs])

        if endpoints:
            for i, l in enumerate(endpoints):
                endpoints[i][0] = proList[l[0]]
                endpoints[i][1] = proList[l[1]]
            endpoints.sort()
            endpointsOuter = []
            n = 0
            for i, point in enumerate(endpoints):
                if point[1] > n:
                    endpointsOuter.append(point)
                    n = point[1]
            endpointsDict = {x[0]:x[1] for x in endpointsOuter}
            return endpointsOuter, len(endpointsOuter), endpointsDict
        
        return None, None, {}

    def argsUpdate(self, lines):
        n = 0 
        for i, line in enumerate(lines):
            if line[0] == "DEF":
                n = len(line) -1
                ARGS.update({(DEF, line[1]) : n})
            elif line[0] == "CALL":
                n = len(line) -1
                ARGS.update({(CALL, line[1]) : n})
                 

    
    def copyDef(self, endpointsOuter, numDefs, rawProgram):
        """
        Creates a new program for function virtual machine
        Copies from program
        """
        DEFS = {}
        if numDefs:
            for i in range(numDefs):
                newProgram = []
                startR = endpointsOuter[i][0]
                endR = endpointsOuter[i][1]
                newProgram.append(rawProgram[startR : endR+1])
                DEFS.update({rawProgram[startR+1] : newProgram[0]})
            return DEFS
        else:
            return None


class PreprocessorFunct:
    def __init__(self):
        pass

    def process(self, rawNewProgram):
        '''
        Processes function program into function virtual machine instructions
        returns:
            newProgram, DEFS, endpoints
        
        newProgram - list of machine instructions for fucntion VM
        DEFS - dictionary with functions and their "names" (order of declaration)
        endpoints - begin and end locations for function declarations
        '''
        self.argsUpdate(rawNewProgram)
        print(ARGS)
        proList = self.proList(rawNewProgram)
        endpoints, numDefs, endpointsDict = self.processDef(rawNewProgram)
        fDEFS = self.copyDef(endpoints, numDefs, rawNewProgram)
        newProgram = self.processLines(rawNewProgram, proList)
        return newProgram, fDEFS, endpointsDict



    def proList(self, newProgram):
        '''
        Makes a list of all the places new lines begin
        '''
        lineSet = set()
        i = 0
        while i != len(newProgram):
            cmd = newProgram[i]
            lineSet.add(i)
            if cmd == DEF:
                i += ARGS[(cmd, newProgram[i+1])]
            elif cmd == CALL:
                i += ARGS[(cmd, newProgram[i+1])]
            else:
                i += ARGS[cmd]  
            i += 1
        
        proList = sorted(list(lineSet))
        return proList
        
    def processLines(self, newProgram, proList):
        '''
        For commands that use line pointers 

        Turns arg from line to program location
        '''
        i = 0
        while i < len(newProgram):
            cmd = newProgram[i]
            if cmd in [GOG, GOE, GOL, GOTO]:
                newProgram[i+1] = proList[newProgram[i+1]]
            if cmd == DEF:
                i += ARGS[(cmd, newProgram[i+1])]
            elif cmd == CALL:
                i += ARGS[(cmd, newProgram[i+1])]
            else:
                i += ARGS[cmd]            
            i += 1
        return newProgram

    def processDef(self, rawNewProgram):
        '''
        Finds locations in the program where declerations of functions begin and end.
        Counts the number of functions 
        Returns list of endpoints,  numDefs
        '''
        defs = []
        enddefs = 0
        endpoints = []
        for i, chr in enumerate(rawNewProgram):
            if chr == DEF:
                defs.append(i)
            elif chr == ENDDEF:
                enddefs = i
                endpoints.append([defs.pop(), enddefs])
        endpoints.reverse()
        endpoints.remove(endpoints[0])
        if endpoints:
            endpointsDict = {x[0]:x[1] for x in endpoints}
            return endpoints, len(endpoints), endpointsDict
        return None, None, None

    def argsUpdate(self, rawNewProgram):
        n = 0
        for i, char in enumerate(rawNewProgram):
            if char == DEF:
                n += 2
                name = rawNewProgram[i+1]
                X = rawNewProgram[i+2]
                n += X
                Y = rawNewProgram[i+3+X]
                n += Y + 1
                ARGS.update({(DEF, name): n})
            if char == CALL:
                n += 2
                name = rawNewProgram[i+1]
                X = rawNewProgram[i+1]
                n += X
                ARGS.update({(CALL, name) : n})
        print('TU', ARGS)



    def copyDef(self, endpoints, numDefs, rawNewProgram):
        '''
        Creates a new program for function virtual machine
        Copies from program
        '''
        fDEFS = {}
        if numDefs:
            for i in range(numDefs):
                funcProgram = []
                startR = endpoints[i][0]
                endR = endpoints[i][1]
                funcProgram.append(rawNewProgram[startR : endR+1])
                fDEFS.update({rawNewProgram[startR + 1] : funcProgram[0]})
            return fDEFS
        return None


if __name__ == "__main__":
    newProgram = [420, 0, 1, 1, 1, 2, 100, 5, 3, 301, 5, 1, 3, 100, 1, 2, 400, 100, 0, 2, 400, 421]
    processorFunct = PreprocessorFunct()
    newProgramm, DEFS, endpoints = processorFunct.process(newProgram)
    print(f'PROGRAM: {newProgram}\n>>======================================================\nDEFS: {DEFS}\n>>======================================================\nEndpointsOuter: {endpoints} ')
    




opcodes.py
# I/O: 100
LOAD = 100
PRINT = 101
LOADP = 102
INPUT = 110
INPUTP = 111
COPY = 120


# Arithmetic
INC = 200
ADD = 210
SUB = 220
MUL = 230
DIV = 240

# Jumps
GOG = 300
GOL = 301
GOE = 310
GOTO = 311

# Halt
HALT = 400


#Other
LIST = 500
META = 401
NOP = 402
DEF = 420
ENDDEF = 421
CALL = 430

OPCODES = {
    "LOAD": LOAD,
    "PRINT": PRINT,
    "LOADP": LOADP,
    "INPUT": INPUT,
    "INPUTP": INPUTP,
    "COPY": COPY,
    "INC": INC,
    "ADD": ADD,
    "SUB": SUB,
    "MUL": MUL,
    "DIV": DIV,
    "GOG": GOG,
    "GOL": GOL,
    "GOE": GOE,
    "GOTO": GOTO,
    "HALT": HALT,
    "META": META,
    "NOP": NOP,
    "LIST": LIST,
    "DEF" : DEF,
    "ENDDEF" : ENDDEF,
    "CALL" : CALL
}


ARGS = {
    LOAD: 2,
    PRINT: 1,
    LOADP: 2,
    INPUT: 1,
    INPUTP: 1,
    COPY: 2,
    INC: 1,
    ADD: 3,
    SUB: 3,
    MUL: 3,
    DIV: 3,
    GOG: 3,
    GOL: 3,
    GOE: 3,
    GOTO: 1,
    HALT: 0,
    META: 2,
    NOP: 0,
    LIST: 2,
    ENDDEF: 0,
}
